### PR TITLE
[Feature] Introduce continuous offset for pulsar

### LINKED ISSUE #9038: Introduce continuous offset for pulsar
**Is your feature request related to a problem? Please describe.**
Currently, pulsar provides <LedgerId, EntryId> to locate an Entry, and provides `BatchIndex` for the inner message of Entry.
But there is not a  continuous offset for each Message in Entry and we could not use the offset to search an Entry. 

For protocol handler like KOP, the map between kakfa offset and pulsar position is implenment by split 64 bits to three parts which represents ledgerId, entryId and batchIndex seperatelly. This way makes the max range of ledgerId much lower than Long.MAX_VALUE and ledgerId space will used up quickly, see https://github.com/streamnative/kop/issues/175.

**Describe the solution you'd like**
We should introduce  continuous offset for pulsar, which can be achieved by broker entry metadata introduced in PIP-70.

**Additional context**
We can introduce offset for each message first, and with this feature, we can more things futher 
1. implement a more stable offset manager for KOP 
2. provide more preceise metrics on message level like backlog
3. ...

### PR DESCRIPTION
**Pull request by aloyszhang**
Fixes #9038 
### Motivation

As described in [PIP-70](https://github.com/apache/pulsar/wiki/PIP-70%3A-Introduce-lightweight-broker-entry-metadata).
One of the use case for Broker entry metadata is  providing continuous message sequence-Id for messages in one topic-partition which is useful for Protocol Hanlder like KOP.

This PR enable Pulsar to support continuous offset for message based on Broker entry metadata.

### Modifications

Introduce a new field for broker entry metadta named `offset`;
Introduce a new interceptor type `ManagedLedgerInterceptor` which intercept entry in `ManagedLedger`；
Each partition will be assigned a `ManagedLedgerInterceptor` when `ManagedLedger` created;
Each Entry will be intercept for adding a  monotone increasing offset in Broker entry metadata and the offet is added by batchSize of entry;
Support find position by a given offset.

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 1]
```diff
@@ -550,6 +566,11 @@ public Position addEntry(byte[] data) throws InterruptedException, ManagedLedger
         return addEntry(data, 0, data.length);
     }
 
+    @Override
+    public Position addEntry(byte[] data, int batchSize) throws InterruptedException, ManagedLedgerException {
+        return addEntry(data, batchSize,0, data.length);
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:571]:
```suggestion
        return addEntry(data, batchSize, 0, data.length);
```

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -672,8 +747,27 @@ private synchronized void internalAsyncAddEntry(OpAddEntry addOperation) {
...
+            return true;
+        }
+        try {
+            managedLedgerInterceptor.beforeAddEntry(addOperation, addOperation.getBatchSize());
+            return true;
+        } catch (Exception e) {
+            addOperation.failed(
+                    new ManagedLedgerInterceptException("Interceptor managed ledger before add to bookie failed."));
+            ReferenceCountUtil.release(addOperation.data);
+            log.error("[{}] Failed to interceptor entry before add to bookie.", name, e);
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java:769]:
```suggestion
            log.error("[{}] Failed to intercept adding an entry to bookie.", name, e);
```

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import org.apache.bookkeeper.client.LedgerHandle;
+import org.apache.bookkeeper.mledger.impl.OpAddEntry;
+
+import java.util.Map;
+
+/**
+ * Interceptor for ManagedLedger.
+ * */
+public interface ManagedLedgerInterceptor {
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java:29]: Can you add the interface annotations to the new interface? See https://github.com/apache/pulsar/wiki/PIP-72%3A-Introduce-Pulsar-Interface-Taxonomy%3A-Audience-and-Stability-Classification

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    OpAddEntry beforeAddEntry(OpAddEntry op, int batchSize);
```
Comment (sijie) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java:30]: Can you add javadoc to the methods?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/intercept/ManagedLedgerInterceptorImpl.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.pulsar.common.api.proto.PulsarApi;
+import org.apache.pulsar.common.intercept.AppendOffsetMetadataInterceptor;
+import org.apache.pulsar.common.intercept.BrokerEntryMetadataInterceptor;
+import org.apache.pulsar.common.protocol.Commands;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ManagedLedgerInterceptorImpl implements ManagedLedgerInterceptor {
+    private static final Logger log = LoggerFactory.getLogger(ManagedLedgerInterceptorImpl.class);
+    private static final String OFFSET = "offset";
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/intercept/ManagedLedgerInterceptorImpl.java:37]: kafka `offset` is not a very good term to be used for describing the index of a message within a log stream. Instead, I would suggest calling it `index` or `logIndex` similar to the term used in the Raft algorithm (https://cs.stackexchange.com/questions/97542/raft-algorithm-whats-the-meaning-of-concept-index).

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -357,10 +357,7 @@ public void publishMessage(ByteBuf headersAndPayload, PublishContext publishCont
                 messageDeduplication.isDuplicate(publishContext, headersAndPayload);
         switch (status) {
             case NotDup:
-                // intercept headersAndPayload and add entry metadata
-                if (appendBrokerEntryMetadata(headersAndPayload, publishContext)) {
-                    ledger.asyncAddEntry(headersAndPayload, this, publishContext);
-                }
+                ledger.asyncAddEntry(headersAndPayload, this, publishContext);
```
Comment (sijie) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:360]: How is the `batchSize` passed to `asyncAddEntry`? I failed to see how did you do that in this pull request.

At the same time, I think `batchSize` is not a good term. If I understand this correctly, it should be `numberOfMessages`, correct?

**Review by sijie — CHANGES_REQUESTED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java [hunk 7]
```diff
@@ -520,4 +587,14 @@ void asyncSetProperties(Map<String, String> properties, final AsyncCallbacks.Upd
      * Roll current ledger if it is full
      */
     void rollCurrentLedgerIfFull();
+
+    /**
+     * Find position by sequenceId.
+     * */
+    CompletableFuture<PositionImpl> asyncFindPosition(com.google.common.base.Predicate<Entry> predicate);
```
[the remark this answers is not in the document]
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java:594]: `condition` of `OpFindNewest` is a `com.google.common.base.Predicate` type

[the remark this answers is not in the document]
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: fixed

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 1 above)
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: fixed

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 2 above)
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: fixed

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java (hunk 3 above)
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java:33]: fixed

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java (hunk 4 above)
Comment (aloyszhang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/interceptor/ManagedLedgerInterceptor.java:41]: fixed

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/intercept/ManagedLedgerInterceptorImpl.java (hunk 5 above)
Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/intercept/ManagedLedgerInterceptorImpl.java]: modify offset to index

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 6 above)
Comment (aloyszhang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: fixed

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sijie) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: I'd recommend using java.util.Predicate in the interface. You can just write a wrapper to convert a java.util.Predicate to a guava Predicate.

**Review by sijie — APPROVED** [review]

**Review by codelipenghui — APPROVED** [review]

**Comment by Anonymitaet** [comment]
@aloyszhang thanks for your great work.

Shall the changes be documented to the [Pulsar docs](https://pulsar.apache.org/docs/en/next/standalone/)?

If so, could you please help add the docs accordingly? Then you can ping me to review, thanks

**Pull request merged**